lib/shopify_api/collection_pagination.rb

```diff
@@ -0,0 +1,52 @@
+module ShopifyAPI
+  module CollectionPagination
+
+    def next_page?
+      next_page_info.present?
+    end
+
+    def previous_page?
+      previous_page_info.present?
+    end
+
+    def fetch_next_page
+      fetch_page(next_page_info)
+    end
+
+    def fetch_previous_page
+      fetch_page(previous_page_info)
+    end
+
+    private
+
+    AVAILABLE_IN_VERSION = ShopifyAPI::ApiVersion::Unstable.new
+
+    def fetch_page(page_info)
+      return [] unless page_info
+
+      resource_class.where(original_params.merge(page_info: page_info))
+    end
+
+    def previous_page_info
+      @previous_page_info ||= extract_page_info(pagination_link_headers.previous_link)
+    end
+
+    def next_page_info
+      @next_page_info ||= extract_page_info(pagination_link_headers.next_link)
+    end
+
+    def extract_page_info(link_header)
+      raise NotImplementedError unless ShopifyAPI::Base.api_version >= AVAILABLE_IN_VERSION
+
+      return nil unless link_header.present?
+
+      CGI.parse(link_header.url.query)["page_info"][0]
+    end
+
+    def pagination_link_headers
+      @pagination_link_headers ||= ShopifyAPI::PaginationLinkHeaders.new(
+        ShopifyAPI::Base.connection.response["Link"]
```

Comment:
I don't think I understand why we get a response off the base connection here.

Does that mean this gem is not threadsafe?

Reply:
ActiveResource appears to be threadsafe, and [sets the connection class to be threadsafe](https://github.com/rails/activeresource/blob/64cedae3019f022a301aad301dc3b7984bb3a153/lib/active_resource/base.rb#L324). 

We have a few places where we reference the base connection for response attributes. [Here](https://github.com/Shopify/shopify_api/blob/6600fae8afa69d29680a3f0413311c6d00724644/lib/shopify_api/resources/checkout.rb#L19) and [here](https://github.com/Shopify/shopify_api/blob/6600fae8afa69d29680a3f0413311c6d00724644/lib/shopify_api/limits.rb#L61).

This was added in AR 4.1, and we only support >= 4.1 

Of note, is that [headers](https://github.com/Shopify/shopify_api/blob/6600fae8afa69d29680a3f0413311c6d00724644/lib/shopify_api/resources/base.rb#L31) is explicitly handled as threadsafe in this gem, but that was done [before the threadsafe stuff was added](https://github.com/Shopify/shopify_api/pull/95). I believe this can be removed, which I'll look into after this.